### ISSUE TITLE
Implement closing orders for V12

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#1962]
_Imported from trac ticket 1962.
Created by **kbower**
Opened in trac: Oct 3, 2012, 9:47:33 AM
Last modified in trac: Apr 22, 2014, 11:46:17 AM_
***
See #1063

Extra work involved here with relieving inventory and copying costs. 

***
* kbower attached [Other Cost Structure.docx](https://trac.retailarchitects.com/trac/attachment/ticket/1962/Other Cost Structure.docx) on Jan 29, 2014, 2:44:56 PM

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 22, 2014, 1:54:55 PM:_

Jon,
Do finalized EEXs show in ITM_INV2CST_FINAL?  **_if cfg.gers_version <= '12.1.3.3' _**

To verify I understand: the CST in ITM_INV (and also copied to SO_LN at
reservation??) is the CST of ONLY the raw types?  **_yes and yes_**
Is this duplicated in ITM_INV2CST, or does ITM_INV2CST only show costs *in
addition to* the raw costs?  **_only shows “in addition to” costs._**

Can there be more than one cost marked RAW, I assume?  And if so, they are
summed?

Should I understand that cost is taken from only exactly one of the
following sources (and they are not cumulative)?

   1.*Vendor and Minor Code*
   2.*Store and Minor Code*
   3.*Item Code*
   4.*Item Code and Store*
**_It is correct that the cost is defined by exactly one of the hierarchy._**

Also, should I understand that the above hierarchy is inspected at receipt
time for OVR, but for PO receipts, *at PO write-time and NOT receipt time*?
**_Calculation for “additional costs” happens at creation time. ...  PO is not rechecked at receiving time. _**
That is, for a PO, the cost is inspected when the PO is written, based on
expected receipt date?? and then at receipt-time, costs are copied
from PO_LN2CST (which means the above hierarchy is NOT inspected at receipt
time)? Is that accurate?

___
**UPDATE:** We will write to that table for 'EEX' also if cfg.gers_version <= '12.1.3.3', it appears that is when 'EEX' stopped being written:

Version 12.1.3.3 still writes for EEX:
```
SQL> select count(*), max(TRN_DT), o.ord_tp_cd  from ITM_INV2CST_FINAL c,
SO o where o.DEL_DOC_NUM = c.DEL_DOC_NUM group by  o.ord_tp_cd;  2

  COUNT(*) MAX(TRN_DT)          ORD
---------- -------------------- ---
   1269381 15-FEB-2014 12:43:13 SAL
     40734 14-FEB-2014 13:29:08 SER
      8028 13-FEB-2014 08:00:14 EEX
```

**UPDATE TO UPDATE:** We will never write to that table for 'EEX' since we only actually write once off GERS (GERS triggers handle this when <= 12.1.3.3)

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 28, 2014, 9:26:36 AM:_

_From Jon_:
Sorry, been busy on other projects. 
 
ITM_INV2CST_FINAL is for finalized Sales Orders, not Credit Memos. Credit memos are only referenced in that table using the CRM_QTY field.  The CRM_QTY field is incremented when the credit memo is written against the original sales order and will not allow the CRM_QTY field to exceed the QTY field.  I am not sure if EEX works, as we stopped using EEX as we prefer having a document for incoming and outgoing merchandise.
 
ITM_INV2CST only shows “in addition to” costs.
 
The uniqueness to the ITM_INV and ITM_INV2CST table is the ITM_INV.  If a cost  is different (or attributes are different) or an “in addition to” cost is different (or attributes are different), there will be a different inv_id.
In other words, you should never see an single INV_ID in ITM_INV which INV_STATUS differs or a single INV_ID in ITM_INV2CST where po_cd differs.  You can regularly see a sku with different raw costs and/or different “in addition costs”, in the same store code and company code.
 
Calculation for “additional costs” happens at creation time. 
When a purchase order is created, the call out determines the “additional costs” based off of the effective date and the PO.WR_DT.  The “additional costs” for purchase orders are stored in the PO_LN2CST table.
When the merchandise is received, the PO_LN.UNIT_CST is entered into PO_LN$ACTN_HST, ITM_TRN, and ITM_INV.  The PO_LN2CST table updates the ITM_INV2CST table.  PO is not rechecked at receiving time.
When a transfer is finalized, the system looks to ITM$TRN2CST table to see if “additional costs” should be added to the ITM_INV2CST table.  These are also stored in the ITM_INV2_CST_IST_FINAL table.
When Overing product, the only cost added is the freight (FRT) based on the hierarchy.
 
It is correct that the cost is defined by exactly one of the hierarchy.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 28, 2014, 9:27:08 AM:_

1) Is the SO_LN cost (at reservation of inventory) copied straight from ITM_INV?
2) Does that cost only include "raw" costs?
___
Yes and yes.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jan 29, 2014, 3:02:36 PM:_

Jon,
You mentioned "ITM_INV2CST_FINAL is for finalized Sales Orders, not Credit
Memos."  However, this is Slumberland data showing CRMs are present:

```
  1* select count(*), o.ord_tp_cd  from ITM_INV2CST_FINAL c, SO o where
o.DEL_DOC_NUM = c.DEL_DOC_NUM group by o.ord_tp_cd
ersqa> /

COUNT(*) ORD
-------- ---
 1.2E+07 SAL
   87509 CRM
   53197 SER


And here is data from another retailer:

  COUNT(*) ORD
---------- ---
   2147096 SAL
     31125 CRM
     28106 SER
         1 MDB
         6 MCR
      2484 EEX

6 rows selected.
```

Any ideas?  I guess we follow the same logic regardless of order type?

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 5, 2014, 9:58:03 AM:_

Assuming you upgraded to V12 in Jun 2011, it looks like you are correct:
only SAL/SER have been inserted since then:

```
ersqa> select count(*), max(TRN_DT), o.ord_tp_cd  from ITM_INV2CST_FINAL c,
SO o where o.DEL_DOC_NUM = c.DEL_DOC_NUM group by  o.ord_tp_cd;

COUNT(*) MAX(TRN_D ORD
-------- --------- ---
 1.2E+07 01-FEB-14 SAL
   87509 12-JUN-11 CRM
   53674 31-JAN-14 SER
```

Same with the other retailer:
```
SQL> select count(*), max(TRN_DT), o.ord_tp_cd  from ITM_INV2CST_FINAL c,
SO o where o.DEL_DOC_NUM = c.DEL_DOC_NUM group by  o.ord_tp_cd;

  COUNT(*) MAX(TRN_DT)          ORD
---------- -------------------- ---
   2152323 04-FEB-2014 08:59:10 SAL
     31125 26-JUL-2011 23:41:35 CRM
     28110 16-JAN-2014 15:47:38 SER
         1 26-JUL-2011 23:41:35 MDB
         6 26-JUL-2011 23:41:35 MCR
      2484 26-JUL-2011 23:41:35 EEX
```

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 14, 2014, 9:59:00 AM:_

(Changeset [[5059]](../commit/a45fa88c7f2e06f8b57ccd92ec8f98ad6a1b64a9))
* fix V12 Inventory OVR costing to use hierarchy of productcosts/vendorcosts if productcosts don't exist #mor #mis
 * re [#2653](2653), adds most all the V12 costing framework required to implement closing V12 sales (see #1962)

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 19, 2014, 4:30:34 PM:_

Jon,

1) You mentioned the final, calculated cost is recorded
in ITM_INV2CST_FINAL.  I'm having a hard time finding evidence to support
that.

Here is an example of a QTY 2 line:
```
SQL> select * from ITM_INV2CST_FINAL where del_doc_num = '11293M8WHZA' and
DEL_DOC_LN# = 10;

DEL_DOC_NUM    DEL_DOC_LN# SE_PART_SEQ CST_C        FAC        AMT
 CUBIC_AMT INV_ID            QTY    CRM_QTY        CST RCV_DT
REF_DOC_NUM REF_DOC_LN# TRN_DT               C
-------------- ----------- ----------- ----- ---------- ----------
---------- ---------- ---------- ---------- ---------- --------------------
----------- ----------- -------------------- -
      SEQ# PO_CD             PO_LN#
---------- ------------- ----------
11293M8WHZA             10             DCFRT          0
  092613PDNW          1                    36 26-SEP-2013 00:00:00
09093W2ZPUI          19 09-SEP-2013 16:24:07 C
         3

11293M8WHZA             10             FRT            3
  092613PDNW          1                    36 26-SEP-2013 00:00:00
                05-SEP-2013 11:08:47 S
         1

11293M8WHZA             10             MARK           7
  092613PDNW          1                    36 26-SEP-2013 00:00:00
09093W2ZPUI          19 09-SEP-2013 16:24:07 C
         2

11293M8WHZA             10             FRT            3
  110113EQPE          1                    36 01-NOV-2013 00:00:00
                01-NOV-2013 11:10:05 S
         1 102413172202           4

11293M8WHZA             10             MARK           7
  110113EQPE          1                    36 01-NOV-2013 00:00:00
11253W2UUZL           8 26-NOV-2013 10:24:40 C
         2 102413172202           4

11293M8WHZA             10             DCFRT          0
  110113EQPE          1                    36 01-NOV-2013 00:00:00
11253W2UUZL           8 26-NOV-2013 10:24:40 C
         3 102413172202           4


6 rows selected.
```

Please note that the CST is $36 for all these records, which, despite the
OTHER and XFER costs, is the same original cost from the PO_LN, as seen
here:

```
SQL> select UNIT_CST, REPL_CST, IVC_CST from po_ln where PO_CD =
'102413172202' and ln# = 4;

  UNIT_CST   REPL_CST    IVC_CST
---------- ---------- ----------
        36         36
```

I don't believe ITM_INV2CST_FINAL houses the final costs, only the ability
to calculate the final cost, do you agree?  (If not so, how would GERS
re-create the inventory upon CRM finalization?)

**_Yes you are correct.  There is the ability to calculate the final cost by using the raw cost and the individual “other cost” records. _**

2) I'm wondering why there are thousands of records (nearly ten
thousand) in ITM_INV2CST_FINAL for which the INV_ID still exists in ITM_INV?

Here is a recent example:

```
SQL> select * from ITM_INV2CST_FINAL where DEL_DOC_NUM = '1221369CLCG' and
DEL_DOC_LN# = 1;

DEL_DOC_NUM    DEL_DOC_LN# SE_PART_SEQ CST_C        FAC        AMT
 CUBIC_AMT INV_ID            QTY    CRM_QTY        CST RCV_DT
REF_DOC_NUM REF_DOC_LN# TRN_DT               C
-------------- ----------- ----------- ----- ---------- ----------
---------- ---------- ---------- ---------- ---------- --------------------
----------- ----------- -------------------- -
      SEQ# PO_CD             PO_LN#
---------- ------------- ----------
1221369CLCG              1             FRT        12.35
  111413YRPH          1                   255 14-NOV-2013 00:00:00
                14-NOV-2013 13:26:05 S
         1 102813662213           8


SQL> select * from itm_inv where INV_ID='111413YRPH';

INV_ID     ITM_CD    STOR LOC_CD RCV_DT               DEL_DOC_NUM
 DEL_DOC_LN# SE_PART_SEQ IST_WR_DT            IST_ IST_SE    IST_LN#
NAS_OUT_ID I        QTY    OOC_QTY CRPT_ID_NU
---------- --------- ---- ------ -------------------- --------------
----------- ----------- -------------------- ---- ------ ----------
---------- - ---------- ---------- ----------
CRPT_SER_NUM                CST    UP_CHRG DIV PO_CD             PO_LN#
LST_ACT_DT
-------------------- ---------- ---------- --- ------------- ----------
--------------------
111413YRPH 5208441   W3   W3WHS  14-NOV-2013 00:00:00 12223S9CTHW
   6                                                                    R
       1          1
                             36          0     071113632297X          2
27-JAN-2014 14:10:22

```


It appears GERS just recycled the INV_ID and it isn't related to the first
one (INV_ID isn't very unique?)  But please verify: it is accurate that the
reserved ITM_INV record should be deleted at finalization, at the same time
the ITM_INV2CST_FINAL records are created, correct?


**_Yes, again you are correct.  I do not believe that the INV_ID is checked against the ITM_INV2CST_FINAL or ITM_INV2CST_IST_FINAL tables.  Meaning that once the INV_ID is out of the system, it could be regenerated for a different sku and or quantity.  I am not sure what happens if a piece comes back in via a Credit Memo, and the INV_ID is already in use.  I assume that a new record is created for the Credit Memo piece as well.  I think the functionality is similar to the WMS thinking in that if a piece was shipped out from a DC to a store, but then came back into the DC at a much later date, it would create a new tag ID. _**


Thanks!
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 19, 2014, 4:32:44 PM:_

Waiting for Jon O response to above.

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 19, 2014, 4:32:44 PM:_ **waiting**

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 24, 2014, 12:41:03 PM:_

(Changeset [[5096]](../commit/29e1acaa78778bbd9207ff1116524df6a345f086))
* closes #1962, closing orders for V12 #mis #slu
   * note, still cannot close with RF
   * create a "primary key" (unique) constraint on GERS ITM_INV2CST_FINAL table for sqlalchemy purposes

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 24, 2014, 12:41:03 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 24, 2014, 3:01:59 PM:_

broken on actual gers install, where apparently gers does all of this via triggers.

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 24, 2014, 3:01:59 PM:_ **reopened**

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 24, 2014, 3:14:33 PM:_

(Changeset [[5098]](../commit/951d7bffb155a969d4bac67d578ae170486b0cbc))
* closes #1962, skip redundant logic that gers triggers performs when order status updated to 'F'

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 24, 2014, 3:14:33 PM:_ **closed**
